arch: riscv: add initial support of SMP

katsuster opened the pull request:
This patch adds support of SMP for RISC-V privilege systems.

  - Modify boot sequence to wake up slave cores
  - Add functions for initializing slave environments
  - Modify context switch and interrupt handler to manage multi cores
  - Add IPI interrupt support into machine_timer driver

katsuster committed:
boards: soc: riscv: add QEMU RV32 virt board

This patch add definitions and settings for QEMU RV32 virt machine.
These are minimal supports only for CPU, Memory, CLINT and UART.

QEMU has supported multiple machine types. It seems virt machine
is most suitable for developing and testing SMP mode on rv32/rv64
both architecture using QEMU.

 | Support       | sifive_u | spike | virt  |
 +---------------+----------+-------+-------+
 | SMP           | yes      | yes   | yes   |
 | RV32/64       | 64       | 32/64 | 32/64 |
 | UART input    | yes      | no    | yes   |
 | Real machine? | yes      | no    | no    |

Signed-off-by: Katsuhiro Suzuki <katsuhiro@katsuster.net>

katsuster committed:
arch: riscv: add new type of context switch

This patch adds new style of context switch called arch_switch().

Signed-off-by: Katsuhiro Suzuki <katsuhiro@katsuster.net>

katsuster committed:
arch: riscv: add initial support of SMP

This patch adds support of SMP for RISC-V privilege systems.

  - Modify boot sequence to wake up slave cores
  - Add functions for initializing slave environments
  - Modify context switch and interrupt handler to manage
    multi cores
  - Add IPI interrupt support into machine_timer driver

NOTE:
Current implementation depends on 'A' extension to execute
atomic operation like as atomic_cas.

Signed-off-by: Katsuhiro Suzuki <katsuhiro@katsuster.net>

katsuster commented:
Fix checkpatch's warnings.

cwshu committed:
driver: plic: convert irq_lock to spinlock for SMP support

This patch uses spinlock instead of irq_lock to protect critical
sections from other cores in SMP mode.

Signed-off-by: Jim Shu <cwshu@andestech.com>

cwshu committed:
driver: plic: PLIC driver use per-hart registers to support SMP

Some of plic registers are per-hart. Because current plic driver always
accesses hart0 registers, changed the driver to access the registers of
current hart. Change IRQ enable/disable API to enable the irq of all
harts(cores) in SMP. Also, plic_init() clear all harts registers
instead of only hart0 registers.

Signed-off-by: Jim Shu <cwshu@andestech.com>

cwshu committed:
driver: plic: changes save_irq to per-CPU variable to support SMP

static variable save_irq records current irq number. In SMP, each CPU
has it's own current irq number, so use per-CPU variable instead.

Signed-off-by: Jim Shu <cwshu@andestech.com>

katsuster commented:
Fix build error about FPU sharing.

cwshu commented on arch/riscv/include/kernel_arch_func.h:
```diff
@@ -26,18 +26,27 @@ static ALWAYS_INLINE void arch_kernel_init(void)
 static inline bool arch_is_in_isr(void)
 {
-	return _kernel.cpus[0].nested != 0U;
+	return arch_curr_cpu()->nested != 0U;
```
In SMP, accessing per-CPU variable (`nested`) should be protected from CPU changing.
There is a race if the thread accesses other CPU's `nested` variable because it is preempted between getting cpu id and accessing `nested` variable.
Zephyr SMP use `arch_irq_lock()` to protect it to be non-preemptive.

Also see [x86 implementation](https://github.com/zephyrproject-rtos/zephyr/blob/zephyr-v2.4.0/arch/x86/include/kernel_arch_func.h#L21) or #22565

katsuster replied:
Thanks, indeed. I'll fix it.
Maybe there are other parts that need to be protected from preemption...

cwshu replied:
I find IPI boardcast may need protection.
Also, get CPU ID by `_current_cpu->id` will also do the assertion check of preemption by `z_smp_cpu_mobile()`

katsuster commented on drivers/timer/riscv_machine_timer.c:
```diff
@@ -16,15 +16,20 @@
+#define RISCV_MSIP_OTHER(id) (RISCV_MSIP_BASE + (uintptr_t)(id) * 4)
+#define RISCV_MSIP     RISCV_MSIP_OTHER(z_riscv_hart_id())
+#define RISCV_MTIMECMP (RISCV_MTIMECMP_BASE + (uintptr_t)z_riscv_hart_id() * 8)
+#define last_count last_count_mp[z_riscv_hart_id()]
```
Thanks. I'm checking strange k_msleep() behavior, this is cause of an issue.

katsuster commented:
I realized the buildkait cannot check CONFIG_SMP part of this patch.
So need to add #28749 PR patches on this PR and re-run buildkite checks...

katsuster commented:
And it's better to have RV64 boards config and buildkait check with it.
I'll add it.

katsuster committed:
boards: soc: riscv: add QEMU RV64 virt board

This patch add definitions and settings for QEMU RV64 virt machine.
These are minimal supports only for CPU, Memory, CLINT and UART.

QEMU has supported multiple machine types. It seems virt machine
is most suitable for developing and testing SMP mode on rv32/rv64
both architecture using QEMU.

 | Support       | sifive_u | spike | virt  |
 +---------------+----------+-------+-------+
 | SMP           | yes      | yes   | yes   |
 | RV32/64       | 64       | 32/64 | 32/64 |
 | UART input    | yes      | no    | yes   |
 | Real machine? | yes      | no    | no    |

Signed-off-by: Katsuhiro Suzuki <katsuhiro@katsuster.net>

cwshu commented:
Hi @katsuster, I find many buildkite errors are about `DT_N_S_soc_S_serial_10000000_IRQ_IDX_0_VAL_irq`.
I also found this issue. It is because dts binding of PLIC driver doesn't support providing IRQ priority from dts currently.
It occurs when using UART16650 and enabling `CONFIG_UART_INTERRUPT_DRIVEN`.
You can temporary fix it by adding the definition into `dts_fixup.h`

katsuster commented:
Hello @cwshu, thank you for comments. To solve this issue, 
importing your 3 patches of PLIC is better than creating from scratch, I think.

- driver: plic: convert irq_lock to spinlock for preparing SMP support
- driver: plic: PLIC driver use per-hart registers to support SMP
- driver: plic: changes save_irq to per-CPU variable to support SMP

Is it OK to use it?

cwshu commented:
Hi @katsuster ,
It's very welcome to use my PLIC patches, but these patches are about PLIC SMP support rather than DTS issue.                                                 
So these patches may not solve your issue? You still needs dts_fixup.h after importing my patches.

katsuster commented:
Hello @cwshu. Oh, I got misunderstanding of this issue...
I'll add dt_fixup.h for now.

cwshu reviewed (commented):
Also, there is a issue in the unmodified code:
current `soc_interrupt_init()` implementation isn't SMP-capable now, because it wrongly use `irq_lock()` to disable IRQ, which is `z_smp_global_lock()` in SMP.
If enabling `CONFIG_RISCV_SOC_INTERRUPT_INIT` in SMP, current code get crash due to calling `z_smp_global_lock()` when _current is NULL.
It should be replaced by `arch_irq_lock()` to fix the issue.

I have finished reviewing most of patches.
It's a great work and thank you for contribution.

cwshu commented:
I have a little suggestion that we can prevent more HW harts than configured in Zephyr from running in the `reset.c`.                                         
It's an error handling mechanism and also be convenient for testing or showing SMP performance on the HW.
(e.g. Run 1/2/4 cores Zephyr's SMP Pi sample on 4 cores HW.)

katsuster commented:
Hello @cwshu. Thanks a lot for your review.

Thank you for pointing CONFIG_RISCV_SOC_INTERRUPT_INIT issue, it has been fixed in my local tree.

And thanks again other comments. I'll check it and fix.

And also, Now I'm trying to fix priority setting issue of using ns16550 with PLIC.
It's not so difficult to fix PLIC binding (use interrupt-cell = 2 = pair of  "irq" and "priority")
but it affects all dts files using PLIC.

cwshu commented:
Hi @katsuster,
Yes, the complex part of this issue is to migrate all dts files using PLIC. We may need an another PR for it.
In my opinion, we can use dts_fixup.h in this PR, and then send an another PR to fix the DTS binding of PLIC.

katsuster commented:
Hello @cwshu. Yes, I think so too.
Just try my local environment in this time, and will send another PR.

katsuster committed:
dts/bindings: Add binding for qemu,rv32 CPU

This patch add a 'rv32' binding which is QEMU RISC-V 32bit CPU.

Signed-off-by: Katsuhiro Suzuki <katsuhiro@katsuster.net>

katsuster committed:
dts/bindings: Add binding for qemu,rv64 CPU

This patch add a 'rv64' binding which is QEMU RISC-V 64bit CPU.

Signed-off-by: Katsuhiro Suzuki <katsuhiro@katsuster.net>

katsuster commented:
Hmm, tests/subsys/logging/log_core/logging.log_core is failed at qemu_riscv64??

katsuster commented:
In RISCV && 64BIT environment. Log system is facing alignment bug...

A sizeof(log_source_dynamic_data) = 4, address offset of &__log_dynamic_start[] are:
  - &__log_dynamic_start[0]: +0
  - &__log_dynamic_start[1]: +4
  - &__log_dynamic_start[2]: +8

```
struct log_source_dynamic_data {
	uint32_t filters;
};

static inline uint32_t *log_dynamic_filters_get(uint32_t source_id)
{
	return &__log_dynamic_start[source_id].filters;
}
```

But RISCV 64bit gcc/ld places each log_dynamic_log, log_dynamic_os and log_dynamic_test per 8bytes alignment.
So address offset of log_dynamic_* are:
  - &log_dynamic_log: +0
  - &log_dynamic_os: +8
  - &log_dynamic_test: +16

tests/subsys/logging/log_core cannot set correct log filters in RISCV && 64BIT environment, so test is always going to fail...

katsuster committed:
drivers: timer: reduce max cycles of riscv machine timer

If next_timeout() returns INT_MAX and pass it to
z_clock_set_timeout(), and machine goes to freeze.
Bad scenario as follows:

- If an argument int32_t ticks is set large value 0xffffffff,
    ticks = MAX(MIN(ticks - 1, (int32_t)MAX_TICKS), 0);
  replaces it into MAX_TICKS.
- uint32_t cyc will be set near by 0xffffffff
  (this is 0xfffd7280 in 100 ticks per second).
- Add adjustment to cyc, adjustment max value is MAX_CYC.
  (cyc = 0xffff14fd)
- Over 0x80000000 value of uint32_t is considered as negative
  value of int32_t.
    if ((int32_t)(cyc + last_count - now) < MIN_DELAY)
  This condition is always true.
- Because cyc += CYC_PER_TICK will get overflow, driver sets mtimecmp
  near value of current mtime.
  (cyc = 0x00007fc0)
- Next timer interrupt will happen soon after return from interrupt
  handler.
- By repeating these events, machine cannot go to next instruction,
  and it's going to freeze.

Signed-off-by: Katsuhiro Suzuki <katsuhiro@katsuster.net>

katsuster commented:
A RISC-V 32bit QEMU with icount feature has strange behavior...
CPU0 is in set_mtimecmp() function and CPU1 is in busy loop (ex. atomic_cas()), if so:

```
static void set_mtimecmp(uint64_t time)
{
#ifdef CONFIG_64BIT
	*(volatile uint64_t *)RISCV_MTIMECMP = time;
#else
	volatile uint32_t *r = (uint32_t *)RISCV_MTIMECMP;
	r[1] = 0xffffffff;    //(1)
	r[0] = (uint32_t)time;    //(2)
	r[1] = (uint32_t)(time >> 32);    //(3)
#endif
}
```

In this case, CPU0 sometime takes very long time (0.1s) to go through from (1) to (2) and from (2) to (3)...

katsuster commented:
Rebase.

katsuster commented:
Reopened. The speed is very slow but I'm still trying to support SMP for RISC-V...
